src/profile/hooks/useProfileLoadTracker.ts

```diff
@@ -0,0 +1,119 @@
+import { useCallback, useContext, useEffect, useRef } from 'react';
+
+import { ProfileContext, ProfileContextData } from '../context/ProfileContext';
+import parseGraphQLError from '../helpers/parseGraphQLError';
+import { QueryResult } from './useProfileQuery';
+
+export const STATUS_FAILED = -1;
+export const STATUS_WAITING = 0;
+export const STATUS_SHOULD_LOAD = 1;
+export const STATUS_LOADING = 2;
+export const STATUS_SUCCESS = 3;
+
+export type Status =
+  | typeof STATUS_WAITING
+  | typeof STATUS_LOADING
+  | typeof STATUS_FAILED
+  | typeof STATUS_SUCCESS
+  | typeof STATUS_SHOULD_LOAD;
+
+export type useProfileLoaderHookReturnType = {
+  isProfileLoadComplete: () => boolean;
+  hasExistingProfile: () => boolean;
+  didProfileLoadFail: () => boolean;
+  reloadProfile: () => Promise<QueryResult>;
+  getStatus: () => Status;
+  getProfile: () => ReturnType<ProfileContextData['getProfile']>;
+};
+
+export function shouldLoad(currentStatus?: Status): boolean {
+  return currentStatus === STATUS_SHOULD_LOAD;
+}
+
+export function isLoading(currentStatus?: Status): boolean {
+  return currentStatus === STATUS_LOADING;
+}
+
+export function didLoadFail(currentStatus?: Status): boolean {
+  return currentStatus === STATUS_FAILED;
+}
+
+export function isLoadComplete(currentStatus?: Status): boolean {
+  return didLoadFail(currentStatus) || isLoadedSuccessfully(currentStatus);
+}
+
+export function isLoadedSuccessfully(currentStatus?: Status): boolean {
+  return currentStatus === STATUS_SUCCESS;
+}
+
+function canStartLoading(currentStatus?: Status): boolean {
+  return !isLoading(currentStatus) && !isLoadComplete(currentStatus);
+}
+
+export function resolveChangedStatus(
+  currentStatus: Status,
+  profileContext: ProfileContextData
+): Status | undefined {
+  if (isLoadComplete(currentStatus)) {
+    return undefined;
+  }
+
+  if (
+    profileContext.error &&
+    !parseGraphQLError(profileContext.error).isAllowedError
+  ) {
+    return STATUS_FAILED;
+  }
+
+  if (profileContext.data && !isLoadedSuccessfully(currentStatus)) {
+    return STATUS_SUCCESS;
+  }
+
+  if (canStartLoading(currentStatus) && !shouldLoad(currentStatus)) {
+    return STATUS_SHOULD_LOAD;
+  }
+
+  return undefined;
+}
```

Comment:
All these functions are internal implementation details and shouldn't be exported. These are now only used in tests, not in production code. It's better to test the _behaviour_, not _implementation_. So test through the public API.

Reply:
Removed implementation tests and unused exports. Added one behavioural test.

In React Apps behavioural tests are quite hard to implement so they are clear and cover all cases. Doable of course, but faking API requests and responses, hooks, dom, user behaviour etc makes tests confusing to read and it takes time to find out if everything is also faked correctly. And all the fake stuff must also be maintained.

Easier to expose and test "safe" parts like resolveChangedStatus() which just returns new status and does not change internal status.